[MLIR] Add allow Insert/extract slice option to pack/unpack op

jerryyin opened the pull request:
This PR adds default option below. The new options will come as default to true and not change the original lowering behavior of pack and unpack op.
 - lowerPadLikeWithInsertSlice to packOp (with default = true)
 - lowerUnpadLikeWithExtractSlice to unPackOp (with default = true)

The motivation of the PR is finer granular control of the lowering of pack and unpack Ops. This is useful in particular when we want to guarantee that there's no additional insertslice and extractslice that interfere with tiling. With the original lowering pipeline, packOp and unPackOp may be lowered to insertslice and extractslice when the high dimensions are unit dimensions and no transpose is invovled. Under such circumstances, such insert and extract slice ops will block producer/consumer fusion tile + fuse transforms. With this PR, we will be able to disable such lowering path and allow consumer fusion to go through as expected.

jerryyin committed:
[NFC] Add allowInsertSliceLowering to packOp and allowExtractSliceLowering to UnPackOp

jerryyin committed:
This commit add test cases to allowInsertSliceLowering and
allowExtractSliceLowering

nirvedhmeshram reviewed (approved):
LGTM!

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -295,7 +296,7 @@ FailureOr<LowerPackResult> linalg::lowerPack(RewriterBase &rewriter,
       DBGSNL(); DBGS() << "collapsed type: " << collapsed; DBGSNL(););
 
-  if (packOp.isLikePad()) {
+  if (allowInsertSliceLowering && packOp.isLikePad()) {
```
Give where this check is used ... I feels like the name is a bit misleading. Shouldn't this be more like ... `lowerPad`?

Max191 replied:
Maybe something like `lowerPadLikeWithInsertSlice` and `lowerUnpadLikeWithExtractSlice`?

joker-eph reviewed (commented):
Nit maybe, but the fact that you have tests in a NFC change is a bit surprising to me?

banach-space commented on mlir/test/Dialect/Linalg/transform-lower-pack.mlir:
```diff
@@ -96,6 +96,34 @@ module attributes {transform.with_named_sequence} {
+func.func @pack_disallowed_as_pad(%arg0: tensor<129x47x16x16xf32>, %arg1: tensor<1x1x1x1x136x64x16x16xf32>) -> tensor<1x1x1x1x136x64x16x16xf32> {
+  %cst_0 = arith.constant 0.0 : f32
+  // tensor.pack is lowered to tensor.pad + tensor.expand_shape + tensor.insert_slice
+  //      CHECK: %[[PAD:.*]] = tensor.pad %[[SRC]] low[0, 0, 0, 0] high[7, 17, 0, 0]
+  //      CHECK:   : tensor<129x47x16x16xf32> to tensor<136x64x16x16xf32>
```
If there's no `tensor.insert_slice`, how is the output actually generated?

Max191 replied:
Without the tensor.insert_slice it just follows the normal lowering path, which is well tested by other lit tests. I think it is okay for this test to be a simple check that the option kicked in, and we can avoid having more tests to change if there is ever a change to the pattern implementation.

banach-space replied:
I agree that it's unnecessary to generate `CHECK` lines for behavior already covered elsewhere. However, I see this test as a variation of `@pack_as_pad`, which produces:
```mlir
%pad = tensor.pad
%slice = tensor.insert_slice %pad
return %slice
```

What's the actual sequence here? I was hoping the lowerPack docs would clarify: https://github.com/llvm/llvm-project/blob/157d847ba737b4136aeb1d92912f549ea1c96d4c/mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h#L1122-L1124

However, the description is sparse and mentions reshape and transpose, which don’t seem to apply here. Is `PAD` the actual return value in this case?

jerryyin replied:
No it will be pad + reshape + transpose and return value is from transpose. Adding reshape and transpose kind-of lose its point because this unit test is not to test that it has reshape and transpose -that's for a dozen unit test above. This is very specifically focused on not having insert or extract slice. I'd prefer orthogonal test coverage if we must have it.

> Nit maybe, but the fact that you have tests in a NFC change is a bit surprising to me?

I tends to agree with @joker-eph that this doesn't need to come with such a unit test based on:
 - Optional attributes in other transform ops are not tested
 - This is NFC PR and doesn't change any existing

banach-space commented:
> Under such circumstances, such insert and extract slice ops will block producer/consumer fusion tile + fuse transforms. With this PR, we will be able to disable such lowering path and allow consumer fusion to go through as expected.

Why not add tests for that? I don't really know how fusion is tested, but this summary implies that we should be writing tests showing how these things compose? And demonstrating the impact of this new flag.

On a different note, there's `lower{Un}Pack` and there's `DecomposeOuterUnitDimsPackOpPattern`. We should unify these things at some point.

Max191 commented:
> On a different note, there's `lower{Un}Pack` and there's `DecomposeOuterUnitDimsPackOpPattern`. We should unify these things at some point.

I agree it is good to unify and reduce the overall number of patterns, but just want to note that we should make sure we still have control over how the patterns are applied (that's probably what you were suggesting anyway, but just wanted to add that comment).

banach-space commented:
> just want to note that we should make sure we still have the same type of control over how/which patterns are applied

Agreed - we shouldn't bundle things pre-maturely. However, if these patterns effectively duplicate one another ... That's a topic for another PR though ;-)

Max191 commented:
> Why not add tests for that? I don't really know how fusion is tested, but this summary implies that we should be writing tests showing how these things compose? And demonstrating the impact of this new flag.

I'm not really sure where a test like this belongs. If you have a suggestion for where we could put it that would be very helpful!

Here is an example case where this helps:
```mlir
%producer = tensor.pack %src into %dest ...
%consumer_loop = scf.for ... {
  %slice = tensor.extract_slice %producer ...
  %consumer = "some_op" %slice ...
  ...
}
```
After decomposition this becomes one of 2 options:
```
// With the slice lowering:
%pad = tensor.pad %src ...
%producer = tensor.insert_slice %pad into %dest ...
%consumer_loop = scf.for ... {
  %slice = tensor.extract_slice %producer ...
  %consumer = "some_op" %slice ...
  ...
}

// With the non-slice (expand_shape/transpose) lowering\
%pad = tensor.pad %src
%expand = tensor.expand_shape %pad ...
%producer = linalg.transpose ins(%expand ...
%consumer_loop = scf.for ... {
  %slice = tensor.extract_slice %producer ...
  %consumer = "some_op" %slice ...
  ...
}
```
If we then want to run fusion of `%producer` into the `%consumer_loop`, then this would be blocked by the insert_slice in the first (slice) case, but not in the second (non-slice) case.

So if we want to test this specific case, then we would be testing the composition of these 2 transform ops, and I'm not sure where the correct test directory for this is. Alternatively, we can add a test for the fusion case above (which would be expected to fail fusion), but I'm not sure if that's a very useful test.

jerryyin commented:
I see discussions converging here. My next commit is going to be:
 - Get rid of the unit test
 - Rename `allowInsertSliceLowering` to `lowerPadLikeWithInsertSlice`
 - Rename `allowExtractSliceLowering` to `lowerPadLikeWithExtractSlice`

jerryyin committed:
Address review requests

- Renamed allowInsertSliceLowering to lowerPadLikeWithInsertSlice
- Renamed allowExtractSliceLowering to lowerUnpadLikeWithExtractSlice
- Removed the redundant unit test since this is NFC change

This reverts commit 46b72028918f13f8faf7ee474d6da14f15a246ef.

hanhanW commented:
> On a different note, there's lower{Un}Pack and there's DecomposeOuterUnitDimsPackOpPattern. We should unify these things at some point.

It should be much easier than before today. People have filled the gaps in the past one year.

banach-space commented:
> I see discussions converging here. My next commit is going to be:
> 
> * Get rid of the unit test

I would keep the unit test. 

When I asked about the test that you added (and subsequently removed), it just wasn't clear to me what to look for and what the difference with the existing tests should be. The comment from @Max191 makes that very clear:
```mlir
// Current
func.func @pack_as_pad_default() {
  %pad = tensor.pad %src ...
  %producer = tensor.insert_slice %pad into %dest
}

// New
func.func @pack_as_pad_{disable|no|skip}_insert_slice() {
  %pad = tensor.pad %src
  %expand = tensor.expand_shape %pad ...
  %producer = linalg.transpose ins(%expand ..
}
```

**Suggestion 1**
* This PR should include a test that follows the pattern above ^^^. 

I think that the following test would be sufficient and very illustrative (trimmed for brevity - feel free to re-use/expand):

```mlir
// CHECK-LABEL: func.func @pack_disallowed_as_pad(
func.func @pack_disallowed_as_pad(%arg0, %arg1) {

  //      CHECK: %[[PAD:.*]] = tensor.pad %[[SRC]] 
  //  CHECK-NOT: %[[RES:.*]] = tensor.insert_slice %[[PAD]]
  //      CHECK: %[[PAD_EXPANDED:.*]] = tensor.expand_shape %[[PAD]]
  //      CHECK: %[[RES:.*]] = linalg.transpose %[[PAD_EXPANDED]]
  %pack = tensor.pack %arg0 padding_value(%cst_0 : f32) inner_dims_pos = [0, 1, 2, 3] inner_tiles = [136, 64, 16, 16] into %arg1
    : tensor<129x47x16x16xf32> -> tensor<1x1x1x1x136x64x16x16xf32>
  return %pack :  tensor<1x1x1x1x136x64x16x16xf32>
}
```
--- 

IIUC, these are the existing tests that exercise similar path: https://github.com/llvm/llvm-project/blob/1b2c8f104f9c6f26500ab608060bbc6b7f40f5e1/mlir/test/Dialect/Linalg/transform-lower-pack.mlir#L3-L60

Both of these tests are called `func.func @pack`, so it's hard to tell what the difference is? Are they duplicating one another? IIUC, these are not "[pad-like](https://github.com/llvm/llvm-project/blob/1b2c8f104f9c6f26500ab608060bbc6b7f40f5e1/mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp#L298-L327)"? 

**Suggestion 2**
* Update the name of these tests ^^^ with something more descriptive so that it's easier to identify the edge cases being tested.


---

> I'm not really sure where a test like this belongs. If you have a suggestion for where we could put it that would be very helpful!

Sure, happy to help! I see multiple options:

**Option 1:** Add a test file in test/Dialect/Linalg (e.g. "transform-lower-pack-and-fuse.mlir") and test it there. Looks like there's already [transform-tile-and-fuse.mlir](https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Linalg/transform-tile-and-fuse.mlir), so use that as your source of inspiration :)

**Option 2:** Some e2e tests exercise complex compilation pipelines, e.g. [pack-dynamic-inner-tile.mlir](https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/pack-dynamic-inner-tile.mlir). But that's quite a heavy hammer.

**Option 3:** Add a bit more powerful TD Op that would include fusion and various related patterns that create fusion opportunities.

I'd start with **Option 1** and leave **Option 2** and **Option 3** for the future. On a related note, we should start creating sub-directories in test/Dialect/Linalg/ - there's too many tests prefixed "transform-{}.mlir" 😅 

---

Hope this helps and thanks for working on this 🙏🏻 !

jerryyin committed:
Adding test cases to allowInsertSliceLowering and allowExtractSliceLowering

jerryyin committed:
Add test to verify pack/producer unpack/consumer fusion

jerryyin commented:
@banach-space Thanks for the detailed review and suggestions. It's very helpful to see so much details in your review!

For suggestion 1, please see commit 671829ee1c89b1d5ff82c866f7074a969414698f that adds back the unit test with attributes. Since you insist unit test for this PR is needed, I'm also amending the title of the PR to get rid of `[NFC]` as it touched function interfaces. While I can understand those additional `CHECK` after the `CHECK-NOT` adds the readability of the unit test, I still think those CHECKs are duplicated coverage with previous test cases. I'm following your suggestion anyways for this standalone case that I wouldn't have strong opinion anyways.

For suggestion 2, option 1, see commit 3ad8cd64f687afa5cadc57e40aa228d6a587ed03 for your suggested test case to exercise consumer and producer fusion. I picked Option 1, and created the bare minimal example to illustrate the additional fusion that can be done with new attributes.

Note that I am skipping the comments (Like suggestion 2 main body) around the test cases that already exist before the PR, as I'd like this PR to be focused about the new attributes added, and not distracted with additional refactoring that we may discover. I'd have no problem filing additional PRs dedicated for refactoring though.

banach-space commented:
Thanks for the updates! 

I've run out of cycles today, but this will be my priority for Monday morning. Apologies for the delay!

banach-space reviewed (commented):
Thank you for addressing my comments so comprehensively and for splitting your replies into separate commits - makes reviewing so much easier 🙏🏻 

I've left a few minor comments inline, otherwise looks good to me.

> Under such circumstances, such insert and extract slice ops will block producer/consumer fusion tile + fuse transforms.

Somewhat tangential to this PR, but it would be good to document whether this limitation is meant to be lifted at some point?

banach-space commented on mlir/test/Dialect/Linalg/transform-tile-and-fuse-pack-unpack.mlir:
```diff
@@ -0,0 +1,121 @@
+module {
+  // CHECK-label: func @fuse_pack_as_producer
+  // CHECK:       scf.forall {{.*}} {
+  // CHECK:         linalg.transpose
```
Btw, this will basically copy `linalg.transpose` inside `scf.forall` - what makes this beneficial?

jerryyin replied:
I think @Max191 can potentially give you more details if you are interested. The goal here is not to have the producer fusion flow disrupted. So potentially we can chain additional ops before transpose when enabled. Follow-up PR may come up to further the progress on this direction.

jerryyin committed:
Adding additional negative test cases

- Added additional test cases to demonstrate insert/extract slice will
  block producer/consumer fusion
- Readability enahncements

jerryyin commented on mlir/test/Dialect/Linalg/transform-lower-pack.mlir:
```diff
@@ -96,6 +96,34 @@ module attributes {transform.with_named_sequence} {
+// This is same as pack_as_pad but since we explicitly added {lowerPadLikeWithInsertSlice = false}, it should not
+// be lowered to insert_slice.
+// CHECK-LABEL: func.func @pack_disallowed_as_pad(
+func.func @pack_disallowed_as_pad(%arg0: tensor<129x47x16x16xf32>, %arg1: tensor<1x1x1x1x136x64x16x16xf32>) -> tensor<1x1x1x1x136x64x16x16xf32> {
```
Agreed, this rename will make the test cases more readable. I've updated name for bock pack and unpack cases.

jerryyin commented on mlir/test/Dialect/Linalg/transform-tile-and-fuse-pack-unpack.mlir:
Added one new test case each that use the default option and demonstrate fusion cannot occur in such circumstances.

banach-space reviewed (approved):
LGTM % one final minor request.

Thanks for working on this!

jerryyin committed:
Add additional LIT variable

This help to clearly demonstrate the produer fusion in pack case and
consumer fusion in unpack case.

jerryyin commented on mlir/test/Dialect/Linalg/transform-tile-and-fuse-pack-unpack.mlir:
```diff
@@ -0,0 +1,238 @@
+  // CHECK-label: func @fuse_pack_as_producer_blocked_by_insert_slice
+  // CHECK:       tensor.insert_slice
+  // CHECK:       scf.forall {{.*}} {
+  // CHECK:         scf.forall.in_parallel
```
Good point!

This is addressed in the latest commit. I'll leave this review open for a day or two in case you've seen anything else worth mentioning.
